Mathlib/Algebra/Module/LocalizedModule.lean

```diff
@@ -581,6 +581,72 @@ lemma IsLocalizedModule.eq_iff_exists [IsLocalizedModule S f] {x₁ x₂} :
     apply_fun f at h
     simp_rw [f.map_smul_of_tower, Submonoid.smul_def, ← Module.algebraMap_end_apply R R] at h
     exact ((Module.End_isUnit_iff _).mp <| map_units f c).1 h
+
+theorem IsLocalizedModule.of_linearEquiv (e : M' ≃ₗ[R] M'') [hf : IsLocalizedModule S f] :
```

Comment:
IMO this should be proved using something like
```lean
instance IsLocalizedModule.sup {T : Submonoid R} (g : M' →ₗ[R] M'') [hf : IsLocalizedModule S f]
  [hg : IsLocalizedModule T g] :
    IsLocalizedModule (S ⊔ T) (g ∘ₗ f : M →ₗ[R] M'') where
  map_units := by
    suffices : S ⊔ T ≤ (IsUnit.submonoid _).comap (algebraMap R (Module.End R M''))
    · exact fun x ↦ this x.prop
    apply sup_le
    · intros r hr
      refine (Module.End_isUnit_iff _).mpr ⟨?_, ?_⟩
      · intro x y e
        obtain ⟨⟨x', s⟩, hs⟩ := IsLocalizedModule.surj T g x
        obtain ⟨⟨y', t⟩, ht⟩ := IsLocalizedModule.surj T g y
        have hs' := congr_arg (t • r • ·) hs
        have ht' := congr_arg (s • r • ·) ht
        simp only [Module.algebraMap_end_apply, Submonoid.smul_def, smul_comm r] at e hs' ht' hs ht
        rw [e, smul_comm, ht', ← map_smul, ← map_smul, ← map_smul, ← map_smul] at hs'
        have ⟨c, hc⟩ := IsLocalizedModule.exists_of_eq (S := T) hs'
        simp only [← smul_comm r] at hc
        replace hc := ((Module.End_isUnit_iff _).mp (IsLocalizedModule.map_units f ⟨r, hr⟩)).1 hc
        apply_fun g at hc
        simp only [Submonoid.coe_subtype, map_smul, ← hs, ← ht] at hc
        simp only [← Submonoid.smul_def, mul_comm s, ← mul_smul] at hc
        exact ((Module.End_isUnit_iff _).mp (IsLocalizedModule.map_units g (c * (t * s)))).1 hc
      · intros x
        obtain ⟨⟨x', s⟩, hs⟩ := IsLocalizedModule.surj T g x
        obtain ⟨a, rfl⟩ := ((Module.End_isUnit_iff _).mp (map_units f ⟨r, hr⟩)).2 x'
        obtain ⟨b, hb⟩ := ((Module.End_isUnit_iff _).mp (map_units g s)).2 (g a)
        simp only [Module.algebraMap_end_apply, map_smul, ← hb, smul_comm r] at hs
        exact ⟨b, ((Module.End_isUnit_iff _).mp (map_units g s)).1 hs.symm⟩
    · intros r hr
      exact IsLocalizedModule.map_units g ⟨r, hr⟩
  surj' x := by
    obtain ⟨⟨x', t⟩, ht : _ = g x'⟩ := IsLocalizedModule.surj T g x
    obtain ⟨⟨x'', s⟩, hs : _ = f x''⟩ := IsLocalizedModule.surj S f x'
    exact ⟨⟨x'', ⟨s.1 * t.1, Submonoid.mem_sup.mpr ⟨_, s.prop, _, t.prop, rfl⟩⟩⟩,
      by simp [← hs, ← ht, mul_smul, Submonoid.smul_def]⟩
  exists_of_eq h := by
    obtain ⟨a, ha⟩ := exists_of_eq (S := T) (f := g) h
    simp only [Submonoid.smul_def, ← map_smul] at ha
    obtain ⟨b, hb⟩ := exists_of_eq (S := S) ha
    simp only [Submonoid.smul_def, ← mul_smul] at hb
    exact ⟨⟨_, Submonoid.mem_sup.mpr ⟨_, b.prop, _, a.prop, rfl⟩⟩, hb⟩

theorem IsLocalizedModule.linearEquiv (e : M ≃ₗ[R] M')
    (hS : ∀ x : S, IsUnit (algebraMap R (Module.End R M') x)) :
      IsLocalizedModule S (e : M →ₗ[R] M') where
  map_units := hS
  surj' x := ⟨⟨e.symm x, 1⟩, by simp⟩
  exists_of_eq h := ⟨1, by simpa using h⟩
```

Reply:
This is a nice result but note that `T` should be an explicit argument (`S` already is). You can save a few lines by starting the `map_units` proof with
```
    suffices S ⊔ T ≤ (IsUnit.submonoid _).comap (algebraMap R (Module.End R M'')) from (this ·.prop)
    refine sup_le (fun r hr ↦ ?_) fun r hr ↦ IsLocalizedModule.map_units g ⟨r, hr⟩
    refine (Module.End_isUnit_iff _).mpr ⟨fun x y e ↦ ?_, fun x ↦ ?_⟩
```
Moreover, all the `IsLocalizedModule.` namespace are unnecessary in the proof. And I think we should extract `(Module.End_isUnit_iff _).mp (map_units` to a lemma as it's used so often.

Note that your `IsLocalizedModule.linearEquiv` doesn't use injectivity of `e` and conflicts with a lemma with the same name later in the file. I suggest not adding this lemma and refactor the proof this way:
```
instance LinearEquiv.isLocalizedModule_bot (e : M ≃ₗ[R] M') : IsLocalizedModule ⊥ e.toLinearMap where
  map_units := by rintro ⟨_, rfl⟩; simp_rw [map_one, isUnit_one]
  surj' m := ⟨⟨e.symm m, 1⟩, (one_smul _ _).trans (e.right_inv _).symm⟩
  exists_of_eq h := ⟨1, by simp_rw [one_smul]; exact e.injective h⟩

theorem IsLocalizedModule.of_linearEquiv (e : M' ≃ₗ[R] M'') [hf : IsLocalizedModule S f] :
    IsLocalizedModule S (e ∘ₗ f : M →ₗ[R] M'') := by
  convert IsLocalizedModule.sup S f ⊥ e.toLinearMap; rw [sup_bot_eq]

theorem IsLocalizedModule.of_linearEquiv' (e : M'' ≃ₗ[R] M) [hf : IsLocalizedModule S f] :
    IsLocalizedModule S (f ∘ₗ e : M'' →ₗ[R] M') := by
  convert IsLocalizedModule.sup ⊥ e.toLinearMap S f; rw [bot_sup_eq]
```

Reply:
That lemma is stronger than both `LinearEquiv.isLocalizedModule_bot` and `isLocalizedModule_id` though. The `simpa` in `exists_of_eq` probably uses injectivity. That can be called `LinearEquiv.isLocalizedModule`. 

It is possible to move this result down and use all the `IsLocalizedModule.mk'`, `IsLocaliedModule.smul_injective` etc to simplify the proof I think. I was just showing that the claim is true and it is not hard to prove it in lean.

Reply:
It seems that there is still a lot of work to be done in this file but I get the feeling that this PR is already quite packed. Would it be a good idea to keep it as it is (modulo some small fixes) and then open a new PR to add the above improvements.

Reply:
Indeed, your lemma is stronger, but the advantage of `isLocalizedModule_bot` is it can be an instance. Since yours can't be, using LinearEquiv namespace does make it easier to use. We should probably also add a version that assumes `∀ x : S, IsUnit (algebraMap R (Module.End R M) x)` for symmetry.